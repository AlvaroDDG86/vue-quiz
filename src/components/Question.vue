<template>
  <div class="question">
    <div class="question__title">{{ question?.title }}</div>
    <div class="question__options">
      <div
        class="question__option"
        :class="{
          'question__option--selected': selected && option.id === selected.id,
        }"
        v-for="(option, idx) in question?.options"
        :key="idx"
        @click="optionSelectedHandler(option)"
      >
        {{ option.title }}
      </div>
    </div>
    <div class="question__validate">
      <AppButton :disabled="selected === undefined" @click="checkAnswerhandler">
        Check answer
      </AppButton>
    </div>
  </div>
</template>
<script lang="ts">
import { defineComponent, PropType, ref, toRefs, watch } from "vue";
import { Question } from "@/models/question.model";
import { Answer } from "@/models/answer.model";

export default defineComponent({
  name: "Question",
  props: {
    question: {
      type: Object as PropType<Question>,
      require: true,
    },
  },
  setup(props, { emit }) {
    const selected = ref();
    const { question } = toRefs(props);
    watch(
      question,
      (value) => {
        selected.value = value?.answer;
      },
      {
        immediate: true,
      }
    );
    const optionSelectedHandler = (option: Answer) => {
      selected.value = option;
    };
    const checkAnswerhandler = () => {
      emit("check-answer", selected.value);
    };
    return {
      selected,
      optionSelectedHandler,
      checkAnswerhandler,
    };
  },
});
</script>
<style lang="postcss">
.question {
  @apply bg-gray-100 rounded-lg shadow-md m-2 py-8 w-full;
  @apply md:m-4;
}
.question__title {
  @apply font-bold text-sm text-left py-1 px-3 mb-2;
  @apply md:text-xl;
}
.question__options {
  min-width: 4rem;
  @apply w-full grid grid-cols-1 gap-1 px-3;
  @apply sm:grid-cols-2;
}
.question__option {
  min-height: fit-content;
  @apply rounded border-2 border-green-500 bg-white text-xs flex justify-center items-center cursor-pointer duration-300 shadow-md;
  @apply hover:border-green-700 hover:bg-gray-200 hover:shadow-none;
  @apply md:text-sm md:h-12;
}
.question__option--selected {
  @apply bg-green-500 text-white bottom-0;
  @apply hover:bg-green-500 hover:text-white;
}
.question__validate {
  @apply mt-4;
}
</style>
