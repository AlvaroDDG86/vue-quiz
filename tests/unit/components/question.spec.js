import { shallowMount } from "@vue/test-utils";
import Question from "@/components/Question";

describe("Question", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(Question, {
      propsData: {
        question: {
          id: 1,
          title: "Question prueba",
          options: [{ id: 1, title: "Option 1" }],
        },
      },
      global: {
        stubs: {
          fa: true,
          AppButton: true,
        },
      },
    });
  });
  test("Should match snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
  test("should display the title", () => {
    expect(wrapper.find(".question__title").text()).toBe("Question prueba");
  });
  test("should display the options", () => {
    expect(wrapper.find(".question__option").exists()).toBe(true);
  });
  test("should select an option and emit check-answer with selected value", () => {
    wrapper.find(".question__option").trigger("click");
    expect(wrapper.vm.selected.title).toBe("Option 1");
    wrapper.find("app-button-stub").trigger("click");
    expect(wrapper.emitted("check-answer").length).toBe(1);
    console.log(wrapper.emitted("check-answer"));
    expect(wrapper.emitted("check-answer")[0][0]).toEqual({
      id: 1,
      title: "Option 1",
    });
  });
});
