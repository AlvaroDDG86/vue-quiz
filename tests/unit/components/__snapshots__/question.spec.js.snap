// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question Should match snapshot 1`] = `
<div class="question">
  <div class="question__title">Question prueba</div>
  <div class="question__options">
    <div class="question__option">Option 1</div>
  </div>
  <div class="question__validate">
    <app-button-stub disabled="true"></app-button-stub>
  </div>
</div>
`;
